README.md
![Logo](https://github.com/PriyeshPatel/habhub-chase-car-tracker/raw/master/resources/icon.png)

# HabHub Android Chase Car Application
An Android application to relay location data from the device to Habitat as `listener_telemetry`. Data is then passed onto [SpaceNear.Us](http://spacenear.us/tracker/) to be displayed as a chase car.

## Market
This app is available on the [Google Play Store](https://play.google.com/store/apps/details?id=com.pexat.habhub.chasecartracker).

## Authors
Written by Priyesh Patel. Artwork by Daniel Saul. See AUTHORS for more information.

## ChangeLog
See CHANGELOG.md.

## License

### Source code
See LICENSE.

### Artwork
Licensed under [Attribution-NonCommercial-ShareAlike 2.0 UK: England & Wales (CC BY-NC-SA 2.0)](http://creativecommons.org/licenses/by-nc-sa/2.0/uk/legalcode). By Daniel Saul & Priyesh Patel.


CHANGELOG.md
# ChangeLog
## v1.0.0 - Day Month Year
  * Yet to be released


src/com/pexat/net/SimpleHttp.java
package com.pexat.net;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.URI;

import org.apache.http.HttpEntity;
import org.apache.http.HttpResponse;
import org.apache.http.client.ClientProtocolException;
import org.apache.http.client.HttpClient;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.client.methods.HttpPut;
import org.apache.http.impl.client.DefaultHttpClient;

public class SimpleHttp {

	public static String get(String url) {
		BufferedReader reader = null;
		String page = "";

		try {
			HttpClient client = new DefaultHttpClient();
			HttpGet request = new HttpGet();

			request.setURI(new URI(url));

			HttpResponse response = client.execute(request);

			reader = new BufferedReader(new InputStreamReader(response.getEntity().getContent()));

			StringBuffer buffer = new StringBuffer("");
			String line = "";
			String newline = System.getProperty("line.separator");

			while ((line = reader.readLine()) != null) {
				buffer.append(line + newline);
			}

			reader.close();

			page = buffer.toString();
		} catch (Exception e) {
			e.printStackTrace();
		} finally {
			if (reader != null) {
				try {
					reader.close();
				} catch (IOException e) {
					e.printStackTrace();
				}
			}
		}

		return page;
	}

	public static void post(String url, HttpEntity data) {
		HttpClient httpclient = new DefaultHttpClient();
		HttpPost httpost = new HttpPost(url);

		try {
			httpost.setEntity(data);
			httpclient.execute(httpost);
		} catch (ClientProtocolException e) {
			e.printStackTrace();
		} catch (IOException e) {
			e.printStackTrace();
		}
	}

	public static void put(String url, HttpEntity data) {
		HttpClient httpclient = new DefaultHttpClient();
		HttpPut httpput = new HttpPut(url);

		try {
			httpput.setEntity(data);
			httpclient.execute(httpput);
		} catch (ClientProtocolException e) {
			e.printStackTrace();
		} catch (IOException e) {
			e.printStackTrace();
		}
	}

}


src/com/pexat/habhub/chasecartracker/Main.java
package com.pexat.habhub.chasecartracker;

import android.app.Activity;
import android.app.ActivityManager;
import android.app.ActivityManager.RunningServiceInfo;
import android.app.AlertDialog;
import android.app.Dialog;
import android.content.ComponentName;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.ServiceConnection;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.os.Handler;
import android.os.IBinder;
import android.os.Message;
import android.os.Messenger;
import android.os.RemoteException;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.View.OnClickListener;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

public class Main extends Activity {
	// Preferences
	public static final String PREFS_NAME = "ChaseCarTrackerPrefs";

	// Dialogs
	private static final int DIALOG_SET_CALLSIGN = 1;

	// UI Elements
	private TextView txt_latitude, txt_longitude, txt_altitude, txt_speed, txt_callsign, txt_lastupdated;
	private Button btn_toggle;

	// Service
	private Messenger tsOutgoing = null;
	private boolean isBound = false;
	private final Messenger tsIncoming = new Messenger(new TrackerServiceIncomingHandler());

	// Current callsign
	private String callsign;

	/**
	 * Activity Setup
	 */

	@Override
	public void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.main);

		txt_latitude = (TextView) findViewById(R.id.txt_latitude);
		txt_longitude = (TextView) findViewById(R.id.txt_longitude);
		txt_altitude = (TextView) findViewById(R.id.txt_altitude);
		txt_speed = (TextView) findViewById(R.id.txt_speed);
		txt_callsign = (TextView) findViewById(R.id.txt_callsign);
		txt_lastupdated = (TextView) findViewById(R.id.txt_last_updated);
		btn_toggle = (Button) findViewById(R.id.btn_toggle);

		SharedPreferences settings = getSharedPreferences(PREFS_NAME, 0);
		setNewCallsign(settings.getString("callsign", ""));

		if (trackerServiceRunning()) {
			doBindService();
			btn_toggle.setText("Stop Tracker");
		}

		btn_toggle.setOnClickListener(new OnClickListener() {
			public void onClick(View v) {
				if (!isBound) {
					Intent i = new Intent(Main.this, TrackerService.class);
					i.putExtra("callsign", callsign);
					startService(i);
					doBindService();
					btn_toggle.setText("Stop Tracker");
				} else {
					doUnbindService();
					stopService(new Intent(Main.this, TrackerService.class));
					btn_toggle.setText("Start Tracker");
				}
			}
		});
	}

	@Override
	public void onDestroy() {
		super.onDestroy();
		try {
			doUnbindService();
		} catch (Throwable t) {
			t.printStackTrace();
		}
	}

	/**
	 * Service Communications
	 */

	private class TrackerServiceIncomingHandler extends Handler {
		@Override
		public void handleMessage(Message msg) {
			switch (msg.what) {
				case TrackerService.MSG_GPS_DATA :
					Bundle b = msg.getData();
					txt_latitude.setText(b.getString("latitude"));
					txt_longitude.setText(b.getString("longitude"));
					txt_altitude.setText(b.getString("altitude"));
					txt_speed.setText(b.getString("speed"));
					txt_lastupdated.setText(b.getString("time"));
					break;
				default :
					super.handleMessage(msg);
					break;
			}
		}
	}

	private ServiceConnection tsConnection = new ServiceConnection() {
		public void onServiceConnected(ComponentName name, IBinder service) {
			tsOutgoing = new Messenger(service);
			try {
				Message msg = Message.obtain(null, TrackerService.MSG_REGISTER_CLIENT);
				msg.replyTo = tsIncoming;
				tsOutgoing.send(msg);
			} catch (RemoteException e) {
				e.printStackTrace();
			}
		}

		public void onServiceDisconnected(ComponentName name) {
			tsOutgoing = null;
		}
	};

	private void doBindService() {
		bindService(new Intent(this, TrackerService.class), tsConnection, Context.BIND_AUTO_CREATE);
		isBound = true;
	}

	private void doUnbindService() {
		if (!isBound)
			return;

		if (tsOutgoing != null) {
			try {
				Message msg = Message.obtain(null, TrackerService.MSG_DEREGISTER_CLIENT);
				msg.replyTo = tsIncoming;
				tsOutgoing.send(msg);
			} catch (RemoteException e) {
				e.printStackTrace();
			}
		}

		unbindService(tsConnection);
		isBound = false;
	}

	/**
	 * Dialog & Menu
	 */

	@Override
	protected Dialog onCreateDialog(int id) {
		switch (id) {
			case DIALOG_SET_CALLSIGN :
				LayoutInflater inflater = LayoutInflater.from(this);
				final View layout = inflater.inflate(R.layout.callsign_dialog, null);
				final EditText txt_editCallsign = (EditText) layout.findViewById(R.id.txt_callsign_dialog);
				txt_editCallsign.setText(callsign);

				final AlertDialog d = new AlertDialog.Builder(this).setTitle("Set Callsign").setView(layout).setPositiveButton("OK", null).create();

				d.setOnDismissListener(new DialogInterface.OnDismissListener() {
					public void onDismiss(DialogInterface dialog) {
						String nc = txt_editCallsign.getText().toString();
						if (nc.equals("")) {
							txt_editCallsign.setError("Please enter a callsign");
							d.show();
						} else {
							setNewCallsign(nc);
						}
					}
				});

				return d;

			default :
				return super.onCreateDialog(id);
		}
	}

	@Override
	public boolean onCreateOptionsMenu(Menu menu) {
		MenuInflater inflater = getMenuInflater();
		inflater.inflate(R.menu.options, menu);
		return true;
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		switch (item.getItemId()) {
			case R.id.opt_callsign :
				showDialog(DIALOG_SET_CALLSIGN);
				return true;
			case R.id.opt_about :
				startActivity(new Intent(this, About.class));
				return true;
			default :
				return super.onOptionsItemSelected(item);
		}
	}

	/**
	 * Helper Functions
	 */

	// Check if tracker service is running
	private boolean trackerServiceRunning() {
		ActivityManager manager = (ActivityManager) getSystemService(ACTIVITY_SERVICE);
		for (RunningServiceInfo service : manager.getRunningServices(Integer.MAX_VALUE)) {
			if ("com.pexat.habhub.chasecartracker.TrackerService".equals(service.service.getClassName())) {
				return true;
			}
		}
		return false;
	}

	// Set a new callsign
	private void setNewCallsign(String c) {
		if (c.equals("")) {
			showDialog(DIALOG_SET_CALLSIGN);
			return;
		}

		callsign = c;
		txt_callsign.setText(callsign + "_chase");

		SharedPreferences settings = getSharedPreferences(PREFS_NAME, 0);
		SharedPreferences.Editor editor = settings.edit();
		editor.putString("callsign", callsign);
		editor.commit();

		if (isBound && tsOutgoing != null) {
			try {
				Bundle b = new Bundle();
				b.putString("callsign", callsign);
				Message msg = Message.obtain(null, TrackerService.MSG_SET_CALLSIGN);
				msg.setData(b);
				msg.replyTo = tsIncoming;
				tsOutgoing.send(msg);
			} catch (RemoteException e) {
				e.printStackTrace();
			}
		}
	}
}

src/com/pexat/habhub/chasecartracker/About.java
package com.pexat.habhub.chasecartracker;

import android.app.ListActivity;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.TextView;

public class About extends ListActivity {

	String data[][];

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);

		data = new String[][]{
				{"Author", "Priyesh Patel"},
				{"Version", AppInfo.getApplicationVersion(getApplicationContext())},
				{"Device Information", AppInfo.getDevice()},
				{"Device Software", AppInfo.getDeviceSoftware()}
		};

		setListAdapter(new TwoLineArrayAdapter(this, data));
	}

	@Override
	protected void onListItemClick(ListView l, View v, int position, long id) {
		switch (position) {
			case 0 :
				Intent i = new Intent(Intent.ACTION_VIEW, Uri.parse(getString(R.string.author_url)));
				startActivity(i);
				break;
		}
		super.onListItemClick(l, v, position, id);
	}

	public class TwoLineArrayAdapter extends ArrayAdapter<String[]> {
		private int mListItemLayoutResId;

		public TwoLineArrayAdapter(Context context, String[][] ts) {
			this(context, android.R.layout.two_line_list_item, ts);
		}

		public TwoLineArrayAdapter(Context context, int listItemLayoutResourceId, String[][] ts) {
			super(context, listItemLayoutResourceId, ts);
			mListItemLayoutResId = listItemLayoutResourceId;
		}

		public View getView(int position, View convertView, ViewGroup parent) {
			LayoutInflater inflater = (LayoutInflater) getContext().getSystemService(Context.LAYOUT_INFLATER_SERVICE);

			View listItemView = convertView;
			if (null == convertView) {
				listItemView = inflater.inflate(mListItemLayoutResId, parent, false);
			}

			listItemView.setPadding(20, 20, 20, 20);

			TextView lineOneView = (TextView) listItemView.findViewById(android.R.id.text1);
			TextView lineTwoView = (TextView) listItemView.findViewById(android.R.id.text2);

			String[] t = (String[]) getItem(position);
			lineOneView.setText(lineOneText(t));
			lineTwoView.setText(lineTwoText(t));

			return listItemView;
		}

		public String lineOneText(String[] t) {
			return t[0];
		}

		public String lineTwoText(String[] t) {
			return t[1];
		}
	}
}

src/com/pexat/habhub/chasecartracker/TrackerService.java
package com.pexat.habhub.chasecartracker;

import java.util.ArrayList;
import android.app.Notification;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.os.Bundle;
import android.os.Handler;
import android.os.IBinder;
import android.os.Message;
import android.os.Messenger;
import android.os.RemoteException;
import android.text.format.Time;
import android.widget.Toast;

public class TrackerService extends Service {
	// Notifications
	private NotificationManager notificationManager;

	// Bound Clients
	private ArrayList<Messenger> messengerClients = new ArrayList<Messenger>();
	private final Messenger tsMessenger = new Messenger(new IncomingHandler());

	// Message Types
	public static final int MSG_REGISTER_CLIENT = 1;
	public static final int MSG_DEREGISTER_CLIENT = 2;
	public static final int MSG_GPS_DATA = 3;
	public static final int MSG_SET_CALLSIGN = 3;

	// GPS
	private LocationManager locationManager;
	private LocationListener locationListener;
	private Location location;
	private Time time_lastupdated = new Time();

	// Callsign
	private String callsign;

	/**
	 * Service Setup
	 */

	@Override
	public void onCreate() {
		super.onCreate();

		showNotification();

		locationListener = new MainLocationListener();
		locationManager = (LocationManager) getSystemService(Context.LOCATION_SERVICE);
		locationManager.requestLocationUpdates(LocationManager.GPS_PROVIDER, 60000, 0, locationListener);
		location = null;

		if (locationManager.isProviderEnabled(LocationManager.NETWORK_PROVIDER)) {
			locationManager.requestLocationUpdates(LocationManager.NETWORK_PROVIDER, 60000, 0, locationListener);
			Location networkLocation = locationManager.getLastKnownLocation(LocationManager.NETWORK_PROVIDER);
			if (isBetterLocation(networkLocation, location)) {
				location = networkLocation;
				time_lastupdated.setToNow();
			}
		}

		if (!locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER)) {
			Toast.makeText(getApplicationContext(), "Please enable your GPS!", Toast.LENGTH_LONG).show();
		} else {
			Location GPSLocation = locationManager.getLastKnownLocation(LocationManager.GPS_PROVIDER);
			if (isBetterLocation(GPSLocation, location)) {
				location = GPSLocation;
				time_lastupdated.setToNow();
			}
		}
	}

	@Override
	public void onDestroy() {
		super.onDestroy();
		hideNotification();
		locationManager.removeUpdates(locationListener);
	}

	@Override
	public int onStartCommand(Intent intent, int flags, int startId) {
		callsign = intent.getStringExtra("callsign") + "_chase";
		return super.onStartCommand(intent, flags, startId);
	}

	/**
	 * GPS
	 */

	private class MainLocationListener implements LocationListener {
		public void onLocationChanged(Location l) {
			if (!isBetterLocation(l, location))
				return;

			location = l;
			time_lastupdated.setToNow();
			sendDataToUI(location, time_lastupdated);

			new Thread(new Runnable() {
				public void run() {
					Time time = new Time(Time.TIMEZONE_UTC);
					time.setToNow();

					ListenerTelemetry telemetry = new ListenerTelemetry(callsign, location, time, AppInfo.getDevice(), AppInfo.getDeviceSoftware(), AppInfo.getApplication(),
							AppInfo.getApplicationVersion(getApplicationContext()));

					HabitatInterface.sendListenerTelemetry(telemetry);
				}
			}).start();
		}

		public void onProviderDisabled(String p) {
		}

		public void onProviderEnabled(String p) {
		}

		public void onStatusChanged(String p, int s, Bundle e) {
		}
	}

	/**
	 * Service Communications
	 */

	@Override
	public IBinder onBind(Intent intent) {
		return tsMessenger.getBinder();
	}

	private class IncomingHandler extends Handler {
		@Override
		public void handleMessage(Message msg) {
			switch (msg.what) {
				case MSG_REGISTER_CLIENT :
					messengerClients.add(msg.replyTo);
					sendDataToUI(location, time_lastupdated);
					break;
				case MSG_DEREGISTER_CLIENT :
					messengerClients.remove(msg.replyTo);
					break;
				case MSG_SET_CALLSIGN :
					callsign = msg.getData().getString("callsign") + "_chase";
				default :
					super.handleMessage(msg);
					break;
			}
		}
	}

	private void sendDataToUI(Location l, Time t) {
		if (l == null)
			return;

		for (int i = messengerClients.size() - 1; i >= 0; i--) {
			try {
				Bundle b = new Bundle();
				b.putString("latitude", Double.toString((double) Math.round(l.getLatitude() * 1000000) / 1000000));
				b.putString("longitude", Double.toString((double) Math.round(l.getLongitude() * 1000000) / 1000000));
				b.putString("altitude", Long.toString(Math.round(l.getAltitude())));
				b.putString("speed", Float.toString((float) Math.round(((l.getSpeed() / 1000) * 3600) * 100) / 100));
				b.putString("time", t.format("%H:%M:%S"));
				Message msg = Message.obtain(null, MSG_GPS_DATA);
				msg.setData(b);
				messengerClients.get(i).send(msg);
			} catch (RemoteException e) {
				messengerClients.remove(i);
			}
		}
	}

	/**
	 * Show permanent notification in notifications bar.
	 */

	private void showNotification() {
		notificationManager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
		Notification n = new Notification(R.drawable.icon, getText(R.string.service_notification), 0);
		PendingIntent pi = PendingIntent.getActivity(this, 0, new Intent(this, Main.class), Notification.FLAG_ONGOING_EVENT);
		n.flags = Notification.FLAG_ONGOING_EVENT;
		n.setLatestEventInfo(this, getText(R.string.service_notification_title), getText(R.string.service_notification), pi);
		notificationManager.notify(R.string.service_notification, n);
	}

	private void hideNotification() {
		notificationManager.cancel(R.string.service_notification);
	}

	/**
	 * GPS Helper Functions
	 */

	private static final int ONE_MINUTE = 1000 * 60;

	/**
	 * Determines whether one Location reading is better than the current
	 * Location fix
	 * 
	 * @param location
	 *            The new Location that you want to evaluate
	 * @param currentBestLocation
	 *            The current Location fix, to which you want to compare the new
	 *            one
	 */
	protected boolean isBetterLocation(Location location, Location currentBestLocation) {
		if (currentBestLocation == null)
			return true;
		if (location == null)
			return false;

		// Check whether the new location fix is newer or older
		long timeDelta = location.getTime() - currentBestLocation.getTime();
		boolean isSignificantlyNewer = timeDelta > ONE_MINUTE;
		boolean isSignificantlyOlder = timeDelta < -ONE_MINUTE;
		boolean isNewer = timeDelta > 0;

		// If it's been more than two minutes since the current location, use
		// the new location because the user has likely moved
		if (isSignificantlyNewer) {
			return true;
			// If the new location is more than two minutes older, it must be
			// worse
		} else if (isSignificantlyOlder) {
			return false;
		}

		// Check whether the new location fix is more or less accurate
		int accuracyDelta = (int) (location.getAccuracy() - currentBestLocation.getAccuracy());
		boolean isLessAccurate = accuracyDelta > 0;
		boolean isMoreAccurate = accuracyDelta < 0;
		boolean isSignificantlyLessAccurate = accuracyDelta > 200;

		// Check if the old and new location are from the same provider
		boolean isFromSameProvider = isSameProvider(location.getProvider(), currentBestLocation.getProvider());

		// Determine location quality using a combination of timeliness and
		// accuracy
		if (isMoreAccurate) {
			return true;
		} else if (isNewer && !isLessAccurate) {
			return true;
		} else if (isNewer && !isSignificantlyLessAccurate && isFromSameProvider) {
			return true;
		}
		return false;
	}

	// Checks whether two providers are the same
	private boolean isSameProvider(String provider1, String provider2) {
		if (provider1 == null) {
			return provider2 == null;
		}
		return provider1.equals(provider2);
	}
}


src/com/pexat/habhub/chasecartracker/HabitatInterface.java
package com.pexat.habhub.chasecartracker;

import java.io.UnsupportedEncodingException;

import org.apache.http.HttpEntity;
import org.apache.http.entity.StringEntity;
import org.json.JSONException;
import org.json.JSONObject;

import com.pexat.net.SimpleHttp;

public class HabitatInterface {
	private static String _habitat_url = "http://habitat.habhub.org/";
	private static String _habitat_db = "habitat";

	protected static String getUUID() {
		String page = SimpleHttp.get(_habitat_url + "_uuids");

		try {
			JSONObject j = new JSONObject(page);
			page = j.getJSONArray("uuids").getString(0);
		} catch (JSONException e) {
			e.printStackTrace();
			return "";
		}

		return page;
	}

	protected static void sendListenerTelemetry(ListenerTelemetry l) {
		String putURL = _habitat_url + _habitat_db + "/" + HabitatInterface.getUUID();
		StringEntity stringdata = null;
		try {
			stringdata = new StringEntity(l.getJSON());
		} catch (UnsupportedEncodingException e) {
			e.printStackTrace();
		}
		HttpEntity data = (HttpEntity) stringdata;

		SimpleHttp.put(putURL, data);
	}
}


src/com/pexat/habhub/chasecartracker/ListenerTelemetry.java
package com.pexat.habhub.chasecartracker;

import org.json.JSONException;
import org.json.JSONObject;

import android.location.Location;
import android.text.format.Time;

public class ListenerTelemetry {
	public long time_created;
	public long time_uploaded;
	public String callsign;
	public int hour;
	public int minute;
	public int second;
	public double latitude;
	public double longitude;
	public double speed;
	public long altitude;
	public String device;
	public String device_software;
	public String application;
	public String application_version;

	public ListenerTelemetry() {
	}

	public ListenerTelemetry(String c, Location l, Time t, String d, String ds, String a, String av) {
		this.setCallsign(c);
		this.setLocationData(l);
		this.setTimeData(t);
		this.setClient(d, ds, a, av);
	}

	public void setLocationData(Location l) {
		this.latitude = (double) Math.round(l.getLatitude() * 1000000) / 1000000;
		this.longitude = (double) Math.round(l.getLongitude() * 1000000) / 1000000;
		this.speed = (double) Math.round(((l.getSpeed() / 1000) * 3600) * 100) / 100;
		this.altitude = Math.round(l.getAltitude());
	}

	public void setTimeData(Time t) {
		this.time_created = Long.valueOf(t.format("%s"));
		this.time_uploaded = time_created;
		this.hour = t.hour;
		this.minute = t.minute;
		this.second = t.second;
	}

	public void setCallsign(String c) {
		this.callsign = c;
	}

	public void setClient(String d, String ds, String a, String av) {
		this.device = d;
		this.device_software = ds;
		this.application = a;
		this.application_version = av;
	}

	public String getJSON() {
		JSONObject output = new JSONObject();
		JSONObject data = new JSONObject();
		JSONObject time = new JSONObject();
		JSONObject client = new JSONObject();

		try {
			output.put("type", "listener_telemetry");
			output.put("time_created", this.time_created);
			output.put("time_uploaded", this.time_created);

			data.put("latitude", this.latitude);
			data.put("longitude", this.longitude);
			data.put("speed", this.speed);
			data.put("altitude", this.altitude);
			data.put("callsign", this.callsign);

			client.put("device", this.device);
			client.put("device_software", this.device_software);
			client.put("application", this.application);
			client.put("application_version", this.application_version);

			time.put("hour", this.hour);
			time.put("minute", this.minute);
			time.put("second", this.second);

			data.put("client", client);
			data.put("time", time);
			output.put("data", data);
		} catch (JSONException e) {
			e.printStackTrace();
		}

		return output.toString();
	}
}
